### PR TITLE
Migrate Farallon Staging hub to this repository.

### PR DESCRIPTION
**Pull request by yuvipanda**
`kops` produces kubeconfig files we can check-in to the repo,
and use to authenticate to `kops` clusters on AWS. 

We test this by moving the farallon staging hub to be deployed
from here, for #368 

TODO:

- [x] Figure out why kubeconfig from `kops` doesn't work after about 12h
- [x] Fix timeouts when mounting EFS shares
- [x] Fix hub deployment test when used with profiles
- [x] Fix https://github.com/2i2c-org/pilot-hubs/pull/374#discussion_r625460411, which is causing failures because we're mounting the configmap unconditionally

### CODE REVIEW

File: config/hubs/farallon.cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                return Options(
+                    Integer("worker_cores", 2, min=1, max=16, label="Worker Cores"),
+                    Float("worker_memory", 4, min=1, max=32, label="Worker Memory (GiB)"),
+                    String("image", default="pangeo/pangeo-notebook:latest", label="Image"),
+                    handler=option_handler,
+                )
+              c.Backend.cluster_options = cluster_options
+            idle: |
+              # timeout after 30 minutes of inactivity
+              c.KubeClusterConfig.idle_timeout = 1800
```
Comment (damianavila) [inline config/hubs/farallon.cluster.yaml:164]: OK, how did you build this config file?

File: secrets/farallon.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          created_at: "2021-05-03T18:24:30Z"
+          enc: CiQA4OM7eBvDdmSosNwoV1H4lGprhSnF6VzDyRaW6U1D9qi6XHQSSQBy9hCY8tILK+/MkhaLO9jCz3fDyhOiF1lxq+zw7GZF8USAZ1ihXFp29CLO69PQa5ejJqZffdWydRTJxuQ1Gd+pEHzC2KmCrGc=
+    azure_kv: []
+    hc_vault: []
+    age: []
+    lastmodified: "2021-05-03T18:24:31Z"
+    mac: ENC[AES256_GCM,data:lSuWapMSsXuAeeSCKHGRsVRXtmeUpCQVTTKDFYHfFADyWF+t8TgwNvRvIuicK69esHW8GnRb5saakKzmlZjwFJKZ+lU8ADXRIY38uRo/4OEvzKhi71lNHxSqZHG/RIPbFsuPoF+xNpqBmfoGmKhrJTxP7zs5hKz2zwZ5GuUdgec=,iv:yVgiI7pbLWPXl1++ZVhg31ASIZn7FgjU8NbouldqC5s=,tag:HsQXJX5JbLVuB10c4JzMiA==,type:str]
+    pgp: []
+    unencrypted_suffix: _unencrypted
+    version: 3.7.1
```
Comment (damianavila) [inline secrets/farallon.yaml:33]: In the related issue you suggested:

> Yep! For kops I was thinking we can just ship the kubeconfig generated by KUBECONFIG=secrets/farallon.yaml kops export kubecfg farallon-2i2c.k8s.local --admin but there are probably other options too

When I run that command I get a file in `secrets/farallon.yaml`, but that one seems unencrypted... I guess I am missing some sops stuff.

File: config/hubs/farallon.cluster.yaml (hunk 1 above)
Comment (damianavila) [reply config/hubs/farallon.cluster.yaml:164]: I see similar stuff in the pangeo-hub repo:
* https://github.com/2i2c-org/pangeo-hubs/blob/staging/hub/values.yaml
* https://github.com/2i2c-org/pangeo-hubs/tree/staging/deployments/farallon/config
Did you build it from those pieces?

Comment (yuvipanda) [reply config/hubs/farallon.cluster.yaml:164]: Yeah, I mostly copied that and hacked around it by testing on the current farallon-staging setup. The values in pangeo-hubs was further inspired by code in github.com/pangeo-data/pangeo-cloud-federation/

File: secrets/farallon.yaml (hunk 2 above)
Comment (yuvipanda) [reply secrets/farallon.yaml:33]: Yeah, if you run `sops -i -e <path>` it will encrypt it in place

**Comment by yuvipanda** [comment]
Getting closer! 

image


Now at:

```
2021-05-04T18:10:45Z [Warning] MountVolume.SetUp failed for volume "farallon-staging-home-nfs" : mount failed: exit status 32 Mounting command: systemd-run Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/338bdd08-d55c-4a43-acf6-f653bcb87f27/volumes/kubernetes.io~nfs/farallon-staging-home-nfs --scope -- mount -t nfs -o noresvport,retrans=2,rsize=1048576,soft,timeo=600,wsize=1048576 fs-7b129903.efs.us-east-2.amazonaws.com:/homesfarallon-staging /var/lib/kubelet/pods/338bdd08-d55c-4a43-acf6-f653bcb87f27/volumes/kubernetes.io~nfs/farallon-staging-home-nfs Output: Running scope as unit: run-rf210704da9ed4f5bbe7837b331747708.scope mount.nfs: Connection timed out 
```

File: config/hubs/farallon.cluster.yaml [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        enabled: false
+      base-hub:
+        nfsPVC:
+          nfs:
+            # from https://docs.aws.amazon.com/efs/latest/ug/mounting-fs-nfs-mount-settings.html
+            mountOptions:
+            - rsize=1048576
+            - wsize=1048576 
+            - timeo=600
+            - soft # We pick soft over hard, so NFS lockups don't lead to hung processes
```
Comment (damianavila) [inline config/hubs/farallon.cluster.yaml:22]: when using `soft` they recommend `timeo` of 150

Comment (yuvipanda) [reply config/hubs/farallon.cluster.yaml:22]: They recommend *at least* 150, so I figured I'll keep it at the default

Comment (damianavila) [reply config/hubs/farallon.cluster.yaml:22]: Yep, I read too quickly...

**Comment by yuvipanda** [comment]
>  Figure out why kubeconfig from kops doesn't work after about 12h

```
      --admin duration[=18h0m0s]   export a cluster admin user credential with the given lifetime and add it to the cluster context
```
for `kops export kubecfg farallon-2i2c.k8s.local`

Makes sense it expires after 18h!

**Comment by yuvipanda** [comment]
https://github.com/2i2c-org/pilot-hubs/pull/380 fixes the timeouts. Turns out if you don't have a directory named `/homes/farallon-staging` but try to mount it, it'll just error with a timeout.

https://pilot-hubs.2i2c.org/en/latest/topic/storage-layer.html has more info about how we do storage and why this matters.

**Comment by damianavila** [comment]
> #380 fixes the timeouts. Turns out if you don't have a directory named /homes/farallon-staging but try to mount it, it'll just error with a timeout.

Great! I was going to suggest looking at mount targets and security groups that sometimes generate timeouts at the time to mount EFS.

File: config/hubs/farallon.cluster.yaml (hunk 3 above)
Comment (damianavila) [reply config/hubs/farallon.cluster.yaml:22]: Now that solved the timeout, do you still want to keep the `soft` option?

**Comment by yuvipanda** [comment]
@damianavila for the expiry of the generated kubeconfig, I see two options:

1. Generate an AWS IAM role that can be used to ephemerally generate a kubeconfig, and modify our deploy script to use that
2. Put a large number into ` --admin`

Actually, we can probably do both - do (2) first and follow up with (1)

**Comment by yuvipanda** [comment]
I've done (2), and opened https://github.com/2i2c-org/pilot-hubs/issues/381 to track (1)

**Comment by damianavila** [comment]
> I've done (2), and opened #381 to track (1)

It makes sense to me in the current context.

File: config/hubs/farallon.cluster.yaml (hunk 3 above)
Comment (yuvipanda) [reply config/hubs/farallon.cluster.yaml:22]: Yeah, definitely wanna keep it on soft.

File: hub-templates/daskhub/values.yaml [hunk 4]
```diff
@@ -116,6 +92,29 @@ basehub:
...
+            project_id = cloud_resources['gcp']['projectId']
+
+            release = os.environ['HELM_RELEASE_NAME']
+            bucket_protocol = 'gcs'
+            bucket_name = f'{project_id}-{release}-scratch-bucket'
+            env = {
+              'SCRATCH_BUCKET_PROTOCOL': bucket_protocol,
+              # Matches "daskhub.scratchBUcket.name" helm template
+              'SCRATCH_BUCKET_NAME': bucket_name,
+              'SCRATCH_BUCKET': f'{bucket_protocol}://{bucket_name}',
```
Comment (damianavila) [inline hub-templates/daskhub/values.yaml:113]: From the deleted lines, it seems the `JUPYTERHUB_USER` is needed at the end:

`          value: $(SCRATCH_BUCKET_PROTOCOL)://$(SCRATCH_BUCKET_NAME)/$(JUPYTERHUB_USER)`

Maybe I am missing something here and somehow that is already included in the `bucket_name`?

Comment (yuvipanda) [reply hub-templates/daskhub/values.yaml]: Good catch, @damianavila! Fixed :)

**Review by damianavila — APPROVED** [review]
OK, I think this one is ready!

**Comment by yuvipanda** [comment]
Thanks for all the review, @damianavila!

**Pull request merged**